### PR TITLE
8341483: Clarify special case handling of Types.getArrayType

### PR DESCRIPTION
**Pull request by jddarcy**
Add specification and tests for invalid inputs to getArrayType that generation an exception. As initially written, this PR changes the behavior on NONE and the null type.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8341511](https://bugs.openjdk.org/browse/JDK-8341511) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8341483](https://bugs.openjdk.org/browse/JDK-8341483): Clarify special case handling of Types.getArrayType (**Sub-task** - P4)
 * [JDK-8341511](https://bugs.openjdk.org/browse/JDK-8341511): Clarify special case handling of Types.getArrayType (**CSR**)

### Reviewers
 * [Dan Smith](https://openjdk.org/census#dlsmith) (@dansmithcode - Committer) Review applies to [222cfa97](https://git.openjdk.org/jdk/pull/21346/files/222cfa97aa0a8690436dfcf037fed7c749f514f1)
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21346/head:pull/21346` \
`$ git checkout pull/21346`

Update a local copy of the PR: \
`$ git checkout pull/21346` \
`$ git pull https://git.openjdk.org/jdk.git pull/21346/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21346`

View PR using the GUI difftool: \
`$ git pr show -t 21346`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21346.diff">https://git.openjdk.org/jdk/pull/21346.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21346#issuecomment-2392756393)

### CODE REVIEW

**Comment by jddarcy** [comment]
Different sections of the JLS imply an array of null type is outside of the language:

Per JLS 10.1:

"An array type is written as the name of an element type followed by some number of empty pairs of square brackets []"

Therefore, since the null type is not denotable, in other words from JLS 4.1

"Because the null type has no name, it is impossible to declare a variable of the null type or to cast to the null type. "

it is not possible to declare an array of null type.

**Comment by liach** [comment]
Are array types valid component types of array types? Not sure how it works in the language, but an array type is not a valid component type if it has 255 dimensions.

**Comment by jddarcy** [comment]
> Are array types valid component types of array types? Not sure how it works in the language, but an array type is not a valid component type if it has 255 dimensions.

Yes; an array type is a reference type and a multi-dimensional array is modeled as an array-of-arrays. I don't know if this particular build would check for the dimensional limit, but it wouldn't be an unreasonable case to add to the test.

File: src/java.compiler/share/classes/javax/lang/model/util/Types.java [hunk 1]
```diff
@@ -263,7 +263,9 @@ public interface Types {
      *
      * @param componentType  the component type
      * @throws IllegalArgumentException if the component type is not valid for
-     *          an array, including executable, package, module, and wildcard types
+     *          an array. All valid types are {@linkplain ReferenceType
+     *          reference types} or {@linkplain PrimitiveType primitive types}.
+     *          Invalid types include null, executable, package, module, and wildcard types.
```
Comment (pavelrappo) [inline src/java.compiler/share/classes/javax/lang/model/util/Types.java:268]: Should it be somehow more clear that it's not `null`, but the __Null Type__. Maybe you can link to it as follows?

    {@linkplain NullType null}

**Review by dansmithcode — APPROVED** [review]
Looks good to me

**Review by liach — APPROVED** [review]
Please forgive my derp, that I forgot array types are reference types.

**Review by pavelrappo — APPROVED** [review]

File: src/java.compiler/share/classes/javax/lang/model/util/Types.java (hunk 1 above)
Comment (jddarcy) [reply src/java.compiler/share/classes/javax/lang/model/util/Types.java]: > Should it be somehow more clear that it's not `null`, but the **Null Type**. Maybe you can link to it as follows?
> 
> ```
> {@linkplain NullType null}
> ```

Good suggestion; implemented.

**Review by pavelrappo — APPROVED** [review]

**Review by liach — APPROVED** [review]
The null type was confusing to me too.

**Pull request closed without merging**